docs/CODE_GUIDELINES.md

```diff
@@ -0,0 +1,174 @@
+  Functions such as `forEach()`, `map()`, `filter()`, and many others are
+  already available from ES6/typescript.
+- Do not habitually define concepts one-per-file. Related classes, interfaces,
+  and symbols should live in the same module unless there is an explicit,
+  conscious motivation for separating them.
+- What is a "unit test"? https://www.youtube.com/watch?v=EZ05e7EMOLM
+  - Unit is not "a single class", it's a piece of the software with a clear
+    boundary. It might be a whole microservice, or a chunk of a monolith that
+    is internally consistent.
+- User-facing messages (and log messages):
+  - Put the variable part of a message at the end, preceded by a colon.
+    - `'foo failed: {0}'`
+  - If a message has many variable parts, mention them in the message in
+    a natural way, but surround them with quotes:
+    - `'creating "{0}" in directory: {1}'`
+- Bubble-up error conditions, do not sink them to random hidden places (such as
+  logs only), expecting callers to figure out the failure mode. If a caller
+  spawns a process that fails, the caller should get either an exception or
+  a callback or a return value indicating that some descendant in the call-chain
+  failed, with the failure reason.
+- Refactoring tools allow us to avoid "premature abstraction". Avoid wrappers
+  until the need is clear and obvious.
+  - Example: What is the purpose of `CloudFormation.Resource`, what is the
+    purpose vs SDK types?
+
+## Test guidelines
+
+- Test real codepaths, avoid mocks.
+  - Example 1: instead of mocking various aspects of the `sam local invoke`
+    codepath, assert the string representation of the CLI command generated by
+    `samCliBuild.ts`
+  - Example 2: instead of mocking various aspects of `sam local invoke`
+    codepath, shift all CONFIG decisions earlier in the codepath and test the
+    real values.
+  - Both examples result in faster, more-readable, more-useful tests, and more
+    test-coverage of real codepaths.
+
+## Code guidelines
+
+- Use maps to group variables, rather than ad-hoc/informal group signals. [Example](https://github.com/aws/aws-toolkit-vscode/blob/abed2c4c7e1329da785190e286e567525afa9da5/src/test/shared/utilities/timeoutUtils.test.ts#L107-L113)
+  - PREFER:
+    ```
+    let topic = {
+      var1 = 'foo',
+      var2 = 'bar',
+    }
+    ```
+  - INSTEAD OF:
+    ```
+    let topicVar1 = 'foo'
+    let topicVar2 = 'bar'
+    ```
+- Where formal grouping is not possible, using a common prefix in names helps
+  discoverability and helps other developers understand "these symbols are
+  related".
+  - PREFER:
+    ```
+    export interface AwsSamDebuggerConfigLoose extends AwsSamDebuggerConfig {
+      ...
+    }
+    ```
+  - INSTEAD OF:
+    ```
+    export interface MorePermissiveAwsSamDebuggerConfig extends AwsSamDebuggerConfig {
+      ...
+    }
+    ```
+- Use module-qualified names, to avoid aliasing members.
```

Comment:
Exception is adding a new activation function to the `extension.ts` file (unless you want to reverse that?)

Reply:
We should reverse that